mlir/lib/CAPI/Dialect/Rock.cpp

```diff
@@ -0,0 +1,79 @@
+//===- MIGraphX.cpp - C Interface for MIGraphX dialect---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir-c/Dialect/Rock.h"
+#include "mlir/CAPI/IR.h"
+#include "mlir/CAPI/Pass.h"
+#include "mlir/CAPI/Registration.h"
+#include "mlir/CAPI/Wrap.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/Dialect/Rock/Tuning/ConvContext.h"
+#include "mlir/Dialect/Rock/Tuning/RockTuning.h"
+
+#include <vector>
+
+MLIR_DEFINE_CAPI_DIALECT_REGISTRATION(Rock, rock, mlir::rock::RockDialect)
+DEFINE_C_API_PTR_METHODS(MlirRockTuningSpace, mlir::rock::TunableParams)
+DEFINE_C_API_PTR_METHODS(MlirRockTuningParam, mlir::rock::ParamEntry)
+// DEFINE_C_API_METHOD(MlirRockTuningKey, const void);
+
+using namespace mlir;
+
+MLIR_CAPI_EXPORTED MlirRockTuningSpace
+mlirRockTuningSpaceCreate(MlirModule module) {
+  struct rock::TunableParams *newParams;
+  auto mod = unwrap(module);
+  newParams = rock::createTunableParams(mod);
+  return wrap(newParams);
+}
+
+MLIR_CAPI_EXPORTED int
+mlirRockTuningGetNumParamsQuick(MlirRockTuningSpace params) {
+  auto tuningSpace = unwrap(params);
+  return tuningSpace->numHeuristicQuick;
+}
+
+MLIR_CAPI_EXPORTED
+int mlirRockTuningGetNumParamsFull(MlirRockTuningSpace params) {
+  auto tuningSpace = unwrap(params);
+  return tuningSpace->tuningRange.size();
+}
+
+// Allocate parameter entry to hold string allocation and pass it more easily.
+MLIR_CAPI_EXPORTED MlirRockTuningParam
+mlirRockTuningCreateParamAt(MlirRockTuningSpace params, int pos) {
```

Comment:
Why make this opaque? Why not use a comma splited c string as the current tuning entry?

I think all 3 APIs can be consolidated into 1:
 - mlirRockTuningCreateParamAt
 - mlirRockTuningSetParam
 - mlirRockTuningGetParamStr

Reply:
I strong disagree with having perf_config be a first-class citizen any more than necessary. We want to get rid of it, not entrench it further, IMO.

Reply:
Why not? It is how it is represented in the db, and a c-string is a very straightforward way to represent this information.

Reply:
I'd like to see us constructing `[...]ParamsAttr` as soon as we can inside our code, so that perf_config becomes a very niche serialization format that only exists at the very edges of our API.

This would also mean getting rid of `InitParams[Non]XDL`

Reply:
> I'd like to see us constructing [...]ParamsAttr as soon as we can inside our code

Okay agreed.

> The last one is needed when client wants to get serialized string to write the record back to the DB, so it's needed separately.

Now that I get a better sense of why you want to do this, can you modify the return type of the last function `mlirRockTuningGetParamStr` into a c-string? I see it less beneficial for client to use a `MlirStringRef` instead of having full control of the tuning parameters' life cycle. If they choose to store it and use it later, and if the reference is destroyed, they are run into trouble.

Reply:
Yeah, that sounds better for the user to fully own the data.
I was just concerned about the allocation size but user can use strdup()